### ISSUE TITLE
Is it possible to train \ tune a spacy NER model with "hints" based on rules \ patterns

### ISSUE THREAD

**Issue by ophiry** [explosion/spaCy#3594]
Dummy example:
I want the NER to be able to detect locations, animals and sport groups
a Matcher \ PhraseMatcher\ EntityRuler (*which is more relevant for this use case?*) could be used to add "simple" rules
like: 
locations: Chicago, New York
animals: Bull, Chicken
groups: Chicago Bulls

The NER layer should be able to learn that Chicago Bulls is a group and not a location and animal (like using a matcher alone would give)
and that other combinations of location + animal are sport groups and not location animal pairs (even if the specific combination didn't exists in the training set)

TLDR: I don't want to use the rule based extracted entities as-is, but as hints for another layer that will use them to improve the entity extraction

**Comment by ines** [comment]
Yes, using match patterns to improve statistical models makes a lot of sense and it's actually one of the approaches we're using in our annotation tool [Prodigy](https://prodi.gy) to make it easier to collect training data. The ambiguous examples where rules fail (e.g. "bulls") are actually really interesting, because those are the ones you want the statistical model to handle.

In terms of the pratical implementation, you'd still want your workflow to have two steps:
1. Use matching to extract candidate examples and select the ones you want to use.
2. Train/update the model with the new examples and evaluate it.

Your model wouldn't just queitly update at runtime, because that's not so useful – you usually always want a dedicated training and evaluation step, so you can use some machine learning tricks and also make sure your model is actually improving.

Also don't forget to include entities that the model previously got right – for example, if your sentence with "Chicago Bulls" also includes a person name, you want this to be included in the training data as well. So your workflow coud look like this:

* Use your pattern rules to extract the matches on your texts.
* For each text, also check the `doc.ents` and get the existing entities.
* Combine both and check that they're correct.
* Update your model with those new examples.

The nice thing is that most of this can be automated by a Python script. Also see [this docs section](https://spacy.io/usage/training#training-data) for an example.

> a Matcher \ PhraseMatcher\ EntityRuler (which is more relevant for this use case?)

The `EntityRuler` is basically a more high-level component that uses the `Matcher` and `PhraseMatcher` to find matches in a `Doc`, and automatically adds them to the `doc.ents`. It also supports feeding in large pattern files and serializes them with a model when you save it. It also handles the fact that named entities, by definition, can't overlap, because one token can only be part of one entity. So if you know that what you're looking for are entities, the `EntityRuler` could be more convenient, because it means you'll have to write less code yourself.
